src/tracing/metric_event.h

```diff
@@ -0,0 +1,128 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <matter/tracing/build_config.h>
+#include <tracing/metric_macros.h>
+#include <lib/core/CHIPError.h>
+#include <system/SystemClock.h>
+#include <tracing/metric_keys.h>
+
+namespace chip {
+namespace Tracing {
+
+/**
+ * Define a metric that can be logged. A metric consists of a key-value pair. The value is
+ * currently limited to simple scalar values. The value is interpreted based on the key type.
+ *
+ * Additionally a metric is tagged as either an instant type or marked with a begin and end
+ * for the event. When the latter is used, a duration can be associated between the two events.
+ */
+struct MetricEvent
+{
+    // Metric tag type
+    enum class Tag
+    {
+        Begin,  // Implies tracking a duration
+        End,    // Implies tracking a duration
+        Instant // No duration
+    };
+
+    // Value for the metric
```

Comment:
nit: also redundant. Could we update to explain the code a bit more?

Would std::variant be applicable here to enforce that we always read the correct type (e.g. json does not correctly check types)?

Reply:
I think this design decision would be useful in a comment (like saying this is like a variant however to avoid exception support we use a tagged union)

Reply:
We also use std::variant:

https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/templates/app/cluster-objects-src.zapt#L17

Reply:
I believe we would get exceptions if we access the wrong variant ... however that sounds like a bug.  Could you provide more details why a std::variant is undesirable regarding exceptions? It seems we do want to check for getting the right value.

Reply:
I want this code running in embedded systems. I assume exception handling won't be supported there. Even if you don't access it incorrectly, don't you have the cost to exception handling? Do we use exceptions in the embedded code today ?

Reply:
Is this resolved @andy31415 ?